### PR TITLE
Implement Server Data Notify.

### LINKED ISSUE #452: Certification: RLOC16 TLV should be included in SVR_DATA.ntf for any update on Server's RLOC
Test_7_1_8 in Harness R14 would verify the Router DUT would send revised server data information to the Leader when its child(who is configured as border router)is timeout.

Brief Spec reference:

> 18 server that changes its RLOC16 sends the Leader an SVR_DATA.req with its Prefix TLV
> 19 and/or Service TLV and includes an RLOC16 TLV containing its old RLOC16. Similarly, a
> 20 Parent that times out a child whose RLOC16 appears in the Network Data sends the leader
> 21 an SVR_DATA.req containing only an RLOC16 TLV with the child's RLOC16

Detail reference: Chapter 5.15 (specifically 5.15.6.1)

### PR DESCRIPTION
**Pull request by jwhui**
Resolves #452.

### CODE REVIEW

File: src/core/thread/network_data.cpp [hunk 1]
```diff
@@ -438,5 +446,89 @@ ThreadError NetworkData::Remove(uint8_t *aStart, uint8_t aLength)
...
+    if (aLocal)
+    {
+        ThreadTlv tlv;
+        tlv.SetType(ThreadTlv::kThreadNetworkData);
+        tlv.SetLength(mLength);
+        SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+        SuccessOrExit(error = message->Append(mTlvs, mLength));
+    }
+
+    if (aRloc16 != Mac::kShortAddrInvalid)
```
Comment (librasungirl) [inline src/core/thread/network_data.cpp]: should here be Mac::kShortAddrBroadcast?
> mOldRloc(Mac::kShortAddrBroadcast)

File: src/core/thread/network_data_leader.cpp [hunk 2]
```diff
@@ -470,21 +470,24 @@ void Leader::HandleServerData(void *aContext, Coap::Header &aHeader, Message &aM
...
+    ThreadNetworkDataTlv networkData;
+    ThreadRloc16Tlv rloc16;
 
     otLogInfoNetData("Received network data registration\n");
 
-    aMessage.Read(aMessage.GetOffset(), sizeof(threadNetworkDataTlv), &threadNetworkDataTlv);
-    tlvsLength = threadNetworkDataTlv.GetLength();
+    if (ThreadTlv::GetTlv(aMessage, ThreadTlv::kRloc16, sizeof(rloc16), rloc16) == kThreadError_None)
+    {
+        RemoveRloc(rloc16.GetRloc16());
```
Comment (librasungirl) [inline src/core/thread/network_data_leader.cpp]: If only ThreadRloc16Tlv is included (eg when ED server is removed), when RemoveRloc(), should network data (/stable) version be updated and afterwards DataResponse() be advertised?

File: src/core/thread/network_data_local.cpp [hunk 3]
```diff
@@ -199,79 +195,18 @@ ThreadError Local::UpdateRloc(BorderRouterTlv &aBorderRouter)
...
-    SuccessOrExit(error = message->Append(&threadNetworkDataTlv, sizeof(threadNetworkDataTlv)));
-    SuccessOrExit(error = message->Append(mTlvs, mLength));
-
-    memset(&messageInfo, 0, sizeof(messageInfo));
-    memcpy(&messageInfo.mPeerAddr, &aDestination, sizeof(messageInfo.mPeerAddr));
-    messageInfo.mPeerPort = kCoapUdpPort;
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
-
-    otLogInfoNetData("Sent network data registration\n");
+    SuccessOrExit(error = NetworkData::SendServerDataNotification(true, mOldRloc));
```
Comment (librasungirl) [inline src/core/thread/network_data_local.cpp]: > ThreadRloc16Tlv is used to update the Leader when a server's RLOC16 changes.
If no rloc changes, ThreadRloc16Tlv should not be included in my understanding.

However for two separate `netdataregister`, there would be ThreadRloc16Tlv in the second server data notification.

File: src/core/thread/network_data.hpp [hunk 4]
```diff
@@ -248,8 +250,29 @@ class NetworkData
      */
     int8_t PrefixMatch(const uint8_t *a, const uint8_t *b, uint8_t aLength);
 
+    /**
+     * This method sends a Server Data Notification message to the Leader.
+     *
+     * @param[in]  aDestination  The IPv6 destination.
```
Comment (bukepo) [inline src/core/thread/network_data.hpp]: parameters mismatch.

**Comment by librasungirl** [comment]
@jwhui ,  some inline comments.
Any plan to notify the Leader when the Rloc of Router Border Server changes?

**Comment by jwhui** [comment]
@librasungirl, please review the updated PR.  If 7.1.8 does not check for sending a Server Data Notify when the Border Router's RLOC changes, then I'm inclined to save that for a future PR.

**Comment by librasungirl** [comment]
LGTM 👍

**Pull request merged**